File: Python/SolidDemo.py
from abc import ABC, abstractmethod


# =======================
# 1. SINGLE RESPONSIBILITY PRINCIPLE (SRP)
# =======================

# BAD: One class doing multiple things
class UserBad:
    def __init__(self, name: str = ""):
        self.name = name

    def save_to_db(self) -> None:
        print("Saving user to DB")

    def send_email(self) -> None:
        print("Sending email")


# GOOD: Separate responsibilities
class User:
    def __init__(self, name: str = ""):
        self.name = name


class UserRepository:
    def save(self, user: User) -> None:
        print("Saving user to DB")


class EmailService:
    def send_email(self, user: User) -> None:
        print("Sending email")


# =======================
# 2. OPEN CLOSED PRINCIPLE (OCP)
# =======================

# BAD: Need to modify class for new types
class DiscountCalculatorBad:
    def calculate(self, type_: str) -> float:
        if type_ == "NEW":
            return 10
        elif type_ == "PREMIUM":
            return 20
        elif type_ == "DIWALI":
            return 30
        return 0


# GOOD: Extend without modifying
class DiscountStrategy(ABC):
    @abstractmethod
    def calculate(self) -> float:
        pass


class NewCustomerDiscount(DiscountStrategy):
    def calculate(self) -> float:
        return 10


class PremiumCustomerDiscount(DiscountStrategy):
    def calculate(self) -> float:
        return 20


class DiscountCalculator:
    def calculate(self, strategy: DiscountStrategy) -> float:
        return strategy.calculate()


# =======================
# 3. LISKOV SUBSTITUTION PRINCIPLE (LSP)
# =======================

# BAD: Violates substitution
class BirdBad:
    def fly(self) -> None:
        print("Flying")


class PenguinBad(BirdBad):
    def fly(self) -> None:
        raise NotImplementedError("Can't fly")


# GOOD: Segregate flying behavior into a separate interface
class Bird(ABC):
    pass


class FlyingBird(Bird, ABC):
    @abstractmethod
    def fly(self) -> None:
        pass


class Sparrow(FlyingBird):
    def fly(self) -> None:
        print("Flying")


# =======================
# 4. INTERFACE SEGREGATION PRINCIPLE (ISP)
# =======================

# BAD: Fat interface
class WorkerBad(ABC):
    @abstractmethod
    def work(self) -> None:
        pass

    @abstractmethod
    def eat(self) -> None:
        pass


class RobotBad(WorkerBad):
    def work(self) -> None:
        print("Working")

    def eat(self) -> None:
        raise NotImplementedError("Robot doesn't eat")


# GOOD: Split interfaces
class Workable(ABC):
    @abstractmethod
    def work(self) -> None:
        pass


class Eatable(ABC):
    @abstractmethod
    def eat(self) -> None:
        pass


class Human(Workable, Eatable):
    def work(self) -> None:
        print("Working")

    def eat(self) -> None:
        print("Eating")


class Robot(Workable):
    def work(self) -> None:
        print("Working")


# =======================
# 5. DEPENDENCY INVERSION PRINCIPLE (DIP)
# =======================

# BAD: High-level depends on low-level
class MySQLDatabase:
    def connect(self) -> None:
        print("Connecting to MySQL")


class ApplicationBad:
    def __init__(self):
        self.db = MySQLDatabase()

    def start(self) -> None:
        self.db.connect()


# GOOD: Depend on abstraction
class Database(ABC):
    @abstractmethod
    def connect(self) -> None:
        pass


class MySQL(Database):
    def connect(self) -> None:
        print("Connecting to MySQL")


class PostgreSQL(Database):
    def connect(self) -> None:
        print("Connecting to PostgreSQL")


class Application:
    def __init__(self, db: Database):
        self.db = db

    def start(self) -> None:
        self.db.connect()


# =======================
# MAIN (TESTING)
# =======================
if __name__ == "__main__":
    # SRP
    user = User()
    UserRepository().save(user)
    EmailService().send_email(user)

    # OCP
    calc = DiscountCalculator()
    print(calc.calculate(PremiumCustomerDiscount()))

    # LSP
    bird: FlyingBird = Sparrow()
    bird.fly()

    # ISP
    robot: Workable = Robot()
    robot.work()

    # DIP
    app = Application(MySQL())
    app.start()

File: Golang/SolidDemo.go
package main

import "fmt"

// =======================
// 1. SINGLE RESPONSIBILITY PRINCIPLE (SRP)
// =======================

// BAD: One struct doing multiple things
type UserBad struct {
	Name string
}

func (u *UserBad) SaveToDB() {
	fmt.Println("Saving user to DB")
}

func (u *UserBad) SendEmail() {
	fmt.Println("Sending email")
}

// GOOD: Separate responsibilities
type User struct {
	Name string
}

type UserRepository struct{}

func (r *UserRepository) Save(user *User) {
	fmt.Println("Saving user to DB")
}

type EmailService struct{}

func (e *EmailService) SendEmail(user *User) {
	fmt.Println("Sending email")
}

// =======================
// 2. OPEN CLOSED PRINCIPLE (OCP)
// =======================

// BAD: Need to modify function for new types
type DiscountCalculatorBad struct{}

func (d *DiscountCalculatorBad) Calculate(discountType string) float64 {
	switch discountType {
	case "NEW":
		return 10
	case "PREMIUM":
		return 20
	case "DIWALI":
		return 30
	}
	return 0
}

// GOOD: Extend without modifying
type DiscountStrategy interface {
	Calculate() float64
}

type NewCustomerDiscount struct{}

func (d *NewCustomerDiscount) Calculate() float64 { return 10 }

type PremiumCustomerDiscount struct{}

func (d *PremiumCustomerDiscount) Calculate() float64 { return 20 }

type DiscountCalculator struct{}

func (d *DiscountCalculator) Calculate(strategy DiscountStrategy) float64 {
	return strategy.Calculate()
}

// =======================
// 3. LISKOV SUBSTITUTION PRINCIPLE (LSP)
// =======================

// BAD: Violates substitution
type BirdBad struct{}

func (b *BirdBad) Fly() {
	fmt.Println("Flying")
}

type PenguinBad struct {
	BirdBad
}

func (p *PenguinBad) Fly() {
	panic("Can't fly")
}

// GOOD: Segregate flying behavior into a separate interface
type Bird interface{}

type FlyingBird interface {
	Bird
	Fly()
}

type Sparrow struct{}

func (s *Sparrow) Fly() {
	fmt.Println("Flying")
}

// =======================
// 4. INTERFACE SEGREGATION PRINCIPLE (ISP)
// =======================

// BAD: Fat interface
type WorkerBad interface {
	Work()
	Eat()
}

type RobotBad struct{}

func (r *RobotBad) Work() {
	fmt.Println("Working")
}

func (r *RobotBad) Eat() {
	panic("Robot doesn't eat")
}

// GOOD: Split interfaces
type Workable interface {
	Work()
}

type Eatable interface {
	Eat()
}

type Human struct{}

func (h *Human) Work() {
	fmt.Println("Working")
}

func (h *Human) Eat() {
	fmt.Println("Eating")
}

type Robot struct{}

func (r *Robot) Work() {
	fmt.Println("Working")
}

// =======================
// 5. DEPENDENCY INVERSION PRINCIPLE (DIP)
// =======================

// BAD: High-level depends on low-level
type MySQLDatabase struct{}

func (m *MySQLDatabase) Connect() {
	fmt.Println("Connecting to MySQL")
}

type ApplicationBad struct {
	db *MySQLDatabase
}

func NewApplicationBad() *ApplicationBad {
	return &ApplicationBad{db: &MySQLDatabase{}}
}

func (a *ApplicationBad) Start() {
	a.db.Connect()
}

// GOOD: Depend on abstraction
type Database interface {
	Connect()
}

type MySQL struct{}

func (m *MySQL) Connect() {
	fmt.Println("Connecting to MySQL")
}

type PostgreSQL struct{}

func (p *PostgreSQL) Connect() {
	fmt.Println("Connecting to PostgreSQL")
}

type Application struct {
	db Database
}

func NewApplication(db Database) *Application {
	return &Application{db: db}
}

func (a *Application) SetDB(db Database) {
	a.db = db
}

func (a *Application) Start() {
	a.db.Connect()
}

// =======================
// MAIN (TESTING)
// =======================
func main() {
	// SRP
	user := &User{}
	(&UserRepository{}).Save(user)
	(&EmailService{}).SendEmail(user)

	// OCP
	calc := &DiscountCalculator{}
	fmt.Println(calc.Calculate(&PremiumCustomerDiscount{}))

	// LSP
	var bird FlyingBird = &Sparrow{}
	bird.Fly()

	// ISP
	var robot Workable = &Robot{}
	robot.Work()

	// DIP
	app := NewApplication(&MySQL{})
	app.Start()
}
